File: src/main/java/glass/fred/advent2018/Day1.java
package glass.fred.advent2018;

import com.google.common.io.Resources;
import java.io.IOException;
import java.net.URISyntaxException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

public class Day1 {

    public static void main(String args[]) throws IOException {
        List<String> lines = read("day1.txt");
        System.out.println(part1(lines));
        System.out.println(part2(lines));
    }

    private static List<String> read(String file) throws IOException {
        try {
            return Files.readAllLines(Paths.get(Resources.getResource(file).toURI()));
        } catch (URISyntaxException e) {
            throw new IOException(e);
        }
    }

    private static int part1(List<String> lines) {
        return lines.stream().mapToInt(Integer::valueOf).sum();
    }

    private static int part2(List<String> lines) {
        int[] frequencies = lines.stream().mapToInt(Integer::valueOf).toArray();
        Set<Integer> seen = new HashSet<>();

        int frequency = 0;
        int i = 0;

        do {
            frequency += frequencies[i++ % frequencies.length]; // Repeats frequency list
        } while(seen.add(frequency));

        return frequency;
    }
}

File: src/main/java/glass/fred/advent2019/Day1.kt
package glass.fred.advent2019

import com.google.common.io.Resources
import java.io.File
import kotlin.math.floor

fun main() {
    val input = File(Resources.getResource("2019/day1.txt").toURI())
    val masses = input.readLines().map { it.toInt() }
    println(Day1.part1(masses))
    println(Day1.part2(masses))
}

object Day1 {

    fun part1(masses: List<Int>) = masses.sumBy { fuelRequired(it) }

    private fun fuelRequired(mass: Int) = floor(mass / 3.0).toInt() - 2

    fun part2(masses: List<Int>) = masses.sumBy { recursive(fuelRequired(it)) }

    private fun recursive(value: Int): Int {
        val fuel = fuelRequired(value)
        return if (fuel <= 0) value else value + recursive(fuel)
    }
}